Add `Networks` to `InfrastructureConfig`

defo89 opened the pull request:
# Proposed Changes
- `Infrastructure.spec.networks` defines all networks and their entities in the infra.
- We update `Infrastructure.status.networking.nodes` with `Infrastructure.spec.networks.cidr` which will propagate to `Shoot.status.networking.nodes`.
- Later `Infrastructure.spec.networks.name` can be referenced by `WorkerConfig` to signal IPAM implementation which CIDR to use for IP assignment from pool.

Fixes https://github.com/ironcore-dev/gardener-extension-provider-metal/issues/82

afritzler reviewed (changes requested):
General question: What should happen with this `Infrastructure` configuration? Is somebody creating a `Network` with a given ID? How is the link between the machine configuration and the `Workers` CIDRs configured here? Would it make sense to have the Worker CIDR configured for each worker pool e.g. here: https://github.com/gardener/gardener/blob/master/example/90-shoot.yaml#L66

afritzler commented:
We should only put into the `Infrastructure` type only global configuration for a Shoot. The reference to AWS in #82 is somehow misleading as in the AWS provider you can configure on a VPC and subsequent on an AZ level specific subnet configuration - and then the infrastructure controller creates this VPC configuration. For us as I see it the configuration is always on a Worker Pool level as we need to compute out of that an effective Ignition to boot the machine with. Maybe we should start on a per Worker Pool configuration first until we have a better understanding how we want the general networking configuration to happen? Wdyt?

defo89 committed:
Add `Networks` to `InfrastructureConfig`

To update `Infrastructure.status.networking.nodes` which will propagate to `Shoot.status.networking.nodes`

defo89 commented on pkg/apis/metal/types_infrastructure.go:
```diff
@@ -21,3 +24,13 @@ type InfrastructureConfig struct {
+	// Name is the name for this network.
+	Name string
+	// Workers is the workers subnet range to create.
+	Workers string
```
good point, changed this to `CIDR`

defo89 commented on pkg/apis/metal/types_infrastructure.go:
```diff
@@ -21,3 +24,13 @@ type InfrastructureConfig struct {
+	// Workers is the workers subnet range to create.
+	Workers string
+	// VLAN is the VLAN ID for the workers' subnet.
+	VLAN string
```
also true, now it's `ID`

Nuckal777 reviewed (commented):
I agree with @afritzler. My gut feeling sees the CIDR for workers on the worker pool as well. Does the provider extension contract have a strong opinion here? Edit: Seems so, the [field we're interested in](https://gardener.cloud/docs/gardener/api-reference/extensions/#extensions.gardener.cloud/v1alpha1.InfrastructureStatusNetworking) is part of the `Infrastructure` resource

Would it make sense to validate that the worker pool cidr is part of `.spec.networking.nodes`?

defo89 committed:
use `slices.Equal`

defo89 commented on pkg/controller/infrastructure/actuator_reconcile.go:
```diff
@@ -5,17 +5,56 @@ package infrastructure
+		}
+	}
+
+	if !equalStringSlices(infra.Status.Networking.Nodes, newNodes) {
```
neat, thanks

defo89 commented:
@Nuckal777  **Shoot** `.spec.networking.nodes` has a limitation of single `string`. Using some summary is risky because it may overlap with 1) other non-relevant systems 2) Seed clusters which would break some checks on Gardener side.

Also this field is planned to be [deprecated](https://github.com/gardener/gardener/issues/7582).

To solve this issue for us today, our extension would update **Infrastructure** `status.networking.nodes` which Gardener copies over to **Shoot's** `status.networking.nodes`. 

Since networks and their details (CIDR, ID=vlan id) are already defined in `Infrastructure`, Worker config could just reference the network Name to choose cidr from:
```
    workers:
    - name: worker-bb123
      network:
        name: network-a
        ipFromPool: true  <-- to signal pool name annotation for ServerClaim
```

defo89 commented:
> I had a short alignment with @afritzler and @ScheererJ. The network configuration can stay as is in the `Infrastructure` type, but we would like a optional field on the worker pools, which reference the network the pool is in. This we can pass to the `MachineClass` of the corresponding pool to make it available to IPAM.

Sure, I think this also matches the example 2 messages above. Should we leave the worker pool config to its own PR when we figure out how the IPAM will look like?

Nuckal777 commented:
@afritzler I think you need to give this PR a green stamp.
